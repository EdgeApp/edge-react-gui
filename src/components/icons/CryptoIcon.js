// @flow
import * as React from 'react'
import { StyleSheet, View } from 'react-native'
import FastImage from 'react-native-fast-image'

import { memo, useMemo } from '../../types/reactHooks.js'
import { useSelector } from '../../types/reactRedux.js'
import { getCurrencyIconUris } from '../../util/CdnUris'
import { guessFromCurrencyCode } from '../../util/CurrencyInfoHelpers'
import { fixSides, mapSides, sidesToMargin } from '../../util/sides.js'
import { WalletSyncCircle } from '../progress-indicators/WalletSyncCircle.js'
import { type Theme, cacheStyles, useTheme } from '../services/ThemeContext.js'

type Props = {
  // Main props - If non is specified, would just render an empty view
  walletId?: string, // To allow showing the progress ratio sync circle
  pluginId?: string, // Needed when walletId is not supplied and we still want to get an icon
  tokenId?: string, // Needed when it's a token (not the plugin's native currency)
  // Image props
  mono?: boolean, // To use the mono dark icon logo
  // Styling props
  sizeRem?: number,
  marginRem?: number | number[],
  // Deprecated!!! here for backward compatibility instead of pluginId or tokenId wherever it's not yet easily available
  currencyCode?: string,
  hideSecondary?: boolean // Only show the currency icon for token (no secondary icon for the network)
}

const CryptoIconComponent = (props: Props) => {
  let { pluginId, tokenId, hideSecondary } = props
  const { walletId, mono = false, sizeRem, marginRem, currencyCode } = props

  const theme = useTheme()
  const styles = getStyles(theme)
  const size = theme.rem(sizeRem ?? 2)

  // Track wallets state from account and update the wallet when ready
  const account = useSelector(state => state.core.account)
  const wallet = walletId != null ? account.currencyWallets[walletId] : null
  // If we have a wallet, get the pluginId from it in case it's missing
  if (wallet != null && pluginId == null) pluginId = wallet.currencyInfo.pluginId

  // ---------------------------------------------------------------------
  // HACK to maintain Backward compatibility for now
  // ---------------------------------------------------------------------
  const ids = guessFromCurrencyCode(account, { currencyCode, pluginId, tokenId })
  pluginId = ids.pluginId
  tokenId = ids.tokenId
  // //////////////////////////////////////////////////////////////////////////////// //

  // Primary Currency icon
  const primaryCurrencyIcon = useMemo(() => {
    if (pluginId == null) return null
    // Get Currency Icon URI
    const icon = getCurrencyIconUris(pluginId, tokenId)
    const source = { uri: mono ? icon.symbolImageDarkMono : icon.symbolImage }
    // Return Currency logo from the edge server
    return <FastImage style={StyleSheet.absoluteFill} source={source} />
  }, [pluginId, tokenId, mono])

  // Secondary (parent) currency icon (if it's a token)
  const secondaryCurrencyIcon = useMemo(() => {
    // Return null if no plugin id or not a token
    if (pluginId == null || tokenId == null || tokenId === pluginId) return null
    // Get Parent Icon URI
    const icon = getCurrencyIconUris(pluginId)
    const source = { uri: mono ? icon.symbolImageDarkMono : icon.symbolImage }
    // Return Parent logo from the edge server
    return <FastImage style={styles.parentIcon} source={source} />
  }, [tokenId, pluginId, mono, styles.parentIcon])

  // Main view styling
  const spacingStyle = useMemo(
    () => ({
      ...sidesToMargin(mapSides(fixSides(marginRem, 0), theme.rem)),
      height: size,
      width: size
    }),
    [marginRem, size, theme.rem]
  )

  return (
    <View style={spacingStyle}>
      {wallet != null ? <WalletSyncCircle size={size} wallet={wallet} /> : null}
      {primaryCurrencyIcon}
      {!hideSecondary ? secondaryCurrencyIcon : null}
    </View>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  parentIcon: {
    position: 'absolute',
    bottom: 0,
    right: 0,
    width: '50%',
    height: '50%'
  }
}))

export const CryptoIcon = memo(CryptoIconComponent)
