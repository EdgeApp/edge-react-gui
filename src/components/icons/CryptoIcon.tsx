import * as React from 'react'
import { StyleSheet, View } from 'react-native'
import FastImage from 'react-native-fast-image'

import compromisedIcon from '../../assets/images/compromisedIcon.png'
import { useWatch } from '../../hooks/useWatch'
import { useSelector } from '../../types/reactRedux'
import { getCurrencyIconUris } from '../../util/CdnUris'
import { guessFromCurrencyCode } from '../../util/CurrencyInfoHelpers'
import { fixSides, mapSides, sidesToMargin } from '../../util/sides'
import { WalletSyncCircle } from '../progress-indicators/WalletSyncCircle'
import { cacheStyles, Theme, useTheme } from '../services/ThemeContext'

interface Props {
  // Main props - If non is specified, would just render an empty view
  pluginId?: string // Needed when walletId is not supplied and we still want to get an icon
  tokenId?: string // Needed when it's a token (not the plugin's native currency)
  walletId?: string // To allow showing the progress ratio sync circle

  // Image props
  hideSecondary?: boolean // Only show the currency icon for token (no secondary icon for the network)
  mono?: boolean // To use the mono dark icon logo

  // Styling props
  marginRem?: number | number[]
  sizeRem?: number

  /** @deprecated Provide tokenId instead. */
  currencyCode?: string
}

const CryptoIconComponent = (props: Props) => {
  const { currencyCode, hideSecondary = false, marginRem, mono = false, sizeRem = 2, walletId } = props

  const theme = useTheme()
  const styles = getStyles(theme)
  const size = theme.rem(sizeRem)

  // Track wallets state from account and update the wallet when ready
  const account = useSelector(state => state.core.account)
  const currencyWallets = useWatch(account, 'currencyWallets')
  const wallet = walletId != null ? currencyWallets[walletId] : null
  const compromised = useSelector(state => {
    if (walletId == null) return 0
    const { modalShown = 0 } = state.ui.settings.securityCheckedWallets[walletId] ?? {}
    return modalShown > 0
  })

  // If we have a wallet, get the pluginId from it in case it's missing
  let { pluginId = wallet?.currencyInfo.pluginId, tokenId } = props

  // ---------------------------------------------------------------------
  // HACK to maintain Backward compatibility for now
  // ---------------------------------------------------------------------
  const ids = guessFromCurrencyCode(account, { currencyCode, pluginId, tokenId })
  pluginId = ids.pluginId
  tokenId = ids.tokenId

  // //////////////////////////////////////////////////////////////////////////////// //

  // Primary Currency icon
  const primaryCurrencyIcon = React.useMemo(() => {
    if (pluginId == null) return null

    // Get Currency Icon URI
    const icon = getCurrencyIconUris(pluginId, tokenId)
    const source = { uri: mono ? icon.symbolImageDarkMono : icon.symbolImage }

    // Return Currency logo from the edge server
    return <FastImage style={StyleSheet.absoluteFill} source={source} />
  }, [pluginId, tokenId, mono])

  // Secondary (parent) currency icon (if it's a token)
  const secondaryCurrencyIcon = React.useMemo(() => {
    if (compromised) {
      return <FastImage source={compromisedIcon} style={styles.parentIcon} />
    }

    // Skip if this is not a token:
    if (pluginId == null || tokenId == null || tokenId === pluginId) {
      return null
    }

    // Get Parent Icon URI
    const icon = getCurrencyIconUris(pluginId)
    const source = { uri: mono ? icon.symbolImageDarkMono : icon.symbolImage }

    // Return Parent logo from the edge server
    return <FastImage style={styles.parentIcon} source={source} />
  }, [compromised, mono, pluginId, styles.parentIcon, tokenId])

  // Main view styling
  const spacingStyle = React.useMemo(
    () => ({
      ...sidesToMargin(mapSides(fixSides(marginRem, 0), theme.rem)),
      height: size,
      width: size
    }),
    [marginRem, size, theme.rem]
  )

  return (
    <View style={spacingStyle}>
      {wallet == null ? null : (
        <WalletSyncCircle
          /* key prevents component from being recycled and shared between wallets */
          key={`${wallet.id}${String(tokenId)}`}
          size={size}
          wallet={wallet}
        />
      )}
      {primaryCurrencyIcon}
      {hideSecondary ? null : secondaryCurrencyIcon}
    </View>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  parentIcon: {
    position: 'absolute',
    bottom: 0,
    right: 0,
    width: '50%',
    height: '50%'
  }
}))

export const CryptoIcon = React.memo(CryptoIconComponent)
