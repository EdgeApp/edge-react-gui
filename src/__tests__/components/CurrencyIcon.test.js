/* globals describe beforeEach afterEach it jest expect */
/* eslint-disable flowtype/require-valid-file-annotation */

import * as React from 'react'
import * as reactRedux from 'react-redux'
import ShallowRenderer from 'react-test-renderer/shallow'

import { CryptoIcon } from '../../components/icons/CryptoIcon.js'

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn()
}))

describe('CryptoIcon', () => {
  beforeEach(() => {
    useDispatchMock.mockImplementation(() => () => {})
    useSelectorMock.mockImplementation(selector => selector(mockStore))
  })
  afterEach(() => {
    useDispatchMock.mockClear()
    useSelectorMock.mockClear()
  })

  const useSelectorMock = reactRedux.useSelector
  const useDispatchMock = reactRedux.useDispatch

  const mockStore = {
    core: {
      account: {
        currencyWallets: {
          '332s0ds39f': {
            pluginId: 'bitcoin',
            watch: () => {}
          }
        }
      }
    }
  }

  it('should render with loading props', () => {
    const renderer = new ShallowRenderer()

    const props = {
      pluginId: 'bitcoin',
      tokenId: 'bitcoin',
      walletId: '332s0ds39f',
      size: 5,
      dark: true,
      resizeMode: 'contain',
      marginRem: 1,
      paddingRem: [1, 2]
    }
    const actual = renderer.render(<CryptoIcon {...props} />)

    expect(actual).toMatchSnapshot()
  })
})
