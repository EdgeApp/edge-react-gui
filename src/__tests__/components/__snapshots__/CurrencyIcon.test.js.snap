// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyIcon should render with loading props 1`] = `
<View
  style={
    Object {
      "height": 45,
      "marginBottom": 22,
      "marginLeft": 22,
      "marginRight": 22,
      "marginTop": 22,
      "width": 45,
    }
  }
>
  <WalletSyncCircle
    size={45}
    wallet={
      Object {
        "pluginId": "bitcoin",
        "watch": [Function],
      }
    }
  />
  <FastImage
    source={
      Object {
        "uri": "https://content.edge.app/currencyIcons/bitcoin/bitcoin.png",
      }
    }
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
</View>
`;
